src/Nix/Builtins.hs

```diff
@@ -764,10 +780,12 @@ functionArgs fun = fun >>= \case
 
 toFile :: MonadNix e m => m (NValue m) -> m (NValue m) -> m (NValue m)
 toFile name s = do
-    name' <- fromValue name
+    name' <- fromStringNoContext =<< fromValue name
     s' <- fromValue s
-    mres <- toFile_ (Text.unpack name') (Text.unpack s')
-    toNix $ Text.pack $ unStorePath mres
+    mres <- toFile_ (Text.unpack name') (Text.unpack $ principledStringIgnoreContext s')
```

Comment:
We definitely don't want to drop the context from s'! toFile needs to be able to handle strings that have context, as that context will be turned into the runtime references of the resulting file